lib/spack/llnl/util/filesystem.py

```diff
@@ -2454,26 +2454,73 @@ class WindowsSimulatedRPath:
     and vis versa.
     """
 
-    def __init__(self, package, link_install_prefix=True):
+    def __init__(
+        self,
+        package,
+        base_modification_prefix: Optional[Union[str, pathlib.Path]] = None,
+        link_install_prefix: bool = True,
+    ):
         """
         Args:
             package (spack.package_base.PackageBase): Package requiring links
+            base_modification_prefix (str|pathlib.Path): Path representation indicating
+                the root directory in which to establish the simulated rpath, ie where the
+                symlinks that comprise the "rpath" behavior will be installed.
+
+                Note: This is a mutually exclusive option with `link_install_prefix` using
+                both is an error.
+
+                Default: None
             link_install_prefix (bool): Link against package's own install or stage root.
                 Packages that run their own executables during build and require rpaths to
-                the build directory during build time require this option. Default: install
+                the build directory during build time require this option.
+
+                Default: install
                 root
+
+                Note: This is a mutually exclusive option with `base_modification_prefix`, using
+                both is an error.
         """
         self.pkg = package
-        self._addl_rpaths = set()
+        self._addl_rpaths: set[str] = set()
+        if link_install_prefix and base_modification_prefix:
+            raise RuntimeError(
+                "Invalid combination of arguments given to WindowsSimulated RPath.\n"
+                "Select either `link_install_prefix` to create an install prefix rpath"
+                " or specify a `base_modification_prefix` for any other link type. "
+                "Specifying both arguments is invalid."
+            )
+        if not (link_install_prefix or base_modification_prefix):
+            raise RuntimeError(
+                "Insufficient arguments given to WindowsSimulatedRpath.\n"
+                "WindowsSimulatedRPath requires one of link_install_prefix"
+                " or base_modification_prefix to be specified."
+                " Neither was provided."
+            )
+
         self.link_install_prefix = link_install_prefix
-        self._additional_library_dependents = set()
+        if base_modification_prefix:
+            self._base_modification_prefix = pathlib.Path(base_modification_prefix)
+        self._additional_library_dependents: set[pathlib.Path] = set()
+        if not self.link_install_prefix:
+            tty.debug(f"Generating rpath for non install context: {base_modification_prefix}")
+
+    @property
+    def base_modification_prefix(self) -> pathlib.Path:
+        if self.link_install_prefix:
+            self._base_modification_prefix = pathlib.Path(self.pkg.prefix)
```

Comment:
Can you set this once in the ctor: e.g.

```
if link_install_prefix and base_modification_prefix:
  ... (what was already there)
elif link_install_prefix:
  _base_modification_prefix = ...(self.pkg.prefix)
```

Reply:
Accessing the prefix in the constructor throws an exception as the spec associated with the package instance hasn't been concretized by the time this constructor is invoked. 
This property was introduced as workaround to delay the evaluation.

Reply:
Could you move the definition of `self.win_rpath` to `windows_establish_runtime_linkage` (it no longer needs to be an instance property)?

I overlooked this in the original review, but I think it's best if we guarantee that `WindowsSimulatedRpath` gets packages with concrete specs.

Reply:
Done and pushed!